### DISCUSSION TITLE
Kernel error: DRBG: could not allocate digest TFM handle: hmac(sha512)

### DISCUSSION THREAD

**Discussion by hgl** [strongswan/strongswan#1673] [Q&A] [answered]
I have a strongSwan responder in OpenWrt, when an initiator connects, the responder says the IKE_SA can be established, but then fails with:

```
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[KNL] received netlink error: No such file or directory (2)
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[KNL] unable to add SAD entry with SPI cc059fbf (FAILED)
Wed Apr 26 19:11:49 2023 kern.info kernel: [   30.804063] DRBG: could not allocate digest TFM handle: hmac(sha512)
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[KNL] received netlink error: No such file or directory (2)
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[KNL] unable to add SAD entry with SPI 0d8423c3 (FAILED)
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel
Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[IKE] failed to establish CHILD_SA, keeping IKE_SA
```

I don't think my responder's swanctl.conf uses anything sha512 related:

```
connections {
  example {
    local_addrs = %any
    remote_addrs = %any
    fragmentation = yes
    pools = example
    local {
      auth = pubkey
      id = "example.com"
    }
    remote {
      auth = pubkey
    }
    children {
      guest_child {
        local_ts = 0.0.0.0/0,::/0
        if_id_in = 2
        if_id_out = 2
        esp_proposals = aes128gcm16-ecp256
        mode = tunnel
      }
    version = 2
    mobike = yes
    proposals = aes128gcm16-prfsha256-ecp256
    dpd_delay = 0s
    keyingtries = 3
}
```

I wonder if my kernel missed any modules or my configuration is wrong? If I understand it correctly, charon was trying to use drbg, which I don't think I specified.

These are the plugins I use:

```
loaded plugins: charon nonce pem openssl kernel-netlink socket-default vici
```

These are configure flags:
```
./configure --target=x86_64-openwrt-linux --host=x86_64-openwrt-linux --build=aarch64-unknown-linux-gnu --disable-dependency-tracking --program-prefix="" --program-suffix="" --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --libexecdir=/usr/lib --sysconfdir=/etc --datadir=/usr/share --localstatedir=/var --mandir=/usr/man --infodir=/usr/info --disable-nls  --disable-defaults --enable-ikev2 --enable-ikev1 --enable-load-warning --enable-nonce --enable-mediation --enable-charon   --enable-swanctl --with-random-device=/dev/random --with-urandom-device=/dev/urandom --with-routing-table="220" --with-routing-table-prio="220" --enable-kernel-netlink  --enable-openssl --enable-pem --enable-socket-default --enable-vici ac_cv_search___atomic_load=no
```

Kernel version: `5.15.108`

**Comment by hgl** [comment]
Forgot to mention, strongSwan version is `5.9.10`.

I wonder the error has anything to do with this: https://patchwork.kernel.org/project/linux-crypto/patch/304ee0376383d9ceecddbfd216c035215bbff861.camel@chronox.de/

**Comment by tobiasbrunner** [comment, marked as answer]
> Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
> Wed Apr 26 19:11:49 2023 daemon.info ipsec: 06[KNL] received netlink error: No such file or directory (2)

This already is an error that indicates that kernel modules are missing (possibly support for AES-GCM, but the error could also be a side-effect of something else failing e.g. getting random bytes to initialize the IV generator).



> If I understand it correctly, charon was trying to use drbg, which I don't think I specified.

No, strongSwan doesn't do anything like that (at least not directly). That has to be something the kernel does internally, and again indicates that some crypto modules are missing (in this case support for HMAC-SHA512 that the DRBG uses internally).

> I wonder the error has anything to do with this: https://patchwork.kernel.org/project/linux-crypto/patch/304ee0376383d9ceecddbfd216c035215bbff861.camel@chronox.de/

Sure, if your kernel is missing support for HMAC-SHA512 because it wasn't automatically enabled.

> ac_cv_search___atomic_load=no

Why did you specify this? (There is no configure check for `__atomic_load()`. Well, there was one at some point, but that has been replaced years ago with 5.5.1.)

**Comment by hgl** [comment]
> Sure, if your kernel is missing support for HMAC-SHA512 because it wasn't automatically enabled.

After installing the sha512 module, the issue went away, thanks!

> Why did you specify this?

It was from the strongSwan package's Makefile, I guess it was never removed even if strongSwan no longer needed it. I've sent a PR that removes it. Thanks for the suggestion.
